docs/contributing/unit-test-guideline.md

```diff
@@ -0,0 +1,193 @@
+# Unit Test Guideline
+
+## About
+
+When we create the unit test for each function or package, we should consider test coverage not only code coverage.
+This guideline will help you to create good unit tests.
+
+## Unit Test
+
+Before considering the unit test, we should define what is the unit.
+Unit is a set of procedures or functions, in a procedural or functional language.
+Also, it is a class and its needed classes, in an object or object-oriented language.
+Unit testing validates the basic units of the program in isolation.
+
+## Guideline
+
+### Strategy
+
+Before creating the unit tests, we must know the essentials of strategy.
+
+1. Know what you are testing
+	- Purpose/Consumers of target function or classes or etc / designed by contract defensive program.
+1. Test behaviors and results, not implementation
+	- ensures that tests only fail when there is an actual effect and not due to internal changes.
+1. Test one thing at a time
+	- Each test should have a clear, concise, singular objective.
+1. Make tests readable and understandable
+	- Hold test code to a similar standard as production code.
+1. Make tests deterministic
+	- A test should pass all the time or fail all the time until fixed.
+1. Make tests independent and self-sufficient
+	- Setup, execution, and verification steps in a given test should not depend on running other tests before it. To keep unit tests simple, fast running, and easy to debug, it may be necessary to isolate the class under test.
+1. Repeat yourself when necessary
+	- It is okay to violate the 'do not repeat yourself' principle if it makes tests simpler and easier to read.
+1. Measure code coverage but focus on test coverage
+	- Do not simply to archive code coverage.
+
+### Test case
+
+#### Basic
+
+It is not perfect, but we should try to cover all codes at first.
+To cover all code, the basic way is the creating test cases to complete all branches in the target unit.
+
+Let's see the below function.
+```go
+func calcSum(val ...int32) (sum int32) {
+	if len(val) == 0 {
+		return
+	} else {
+		for _, v := range val {
+			sum += v
+		}
+		return
+	}
+}
+```
+
+When the above function is given, we should create 2 test for achieving 100% code coverage at least.
+
+- When val is not given
+  - In other words, the default value is given
+- When val(int32) is given
+  - `var val int32 = 1`
+  
+It seems enough for the given function, but we have to take care of some test cases that are remaining for improving test coverage.
+In this case, there is one test case is remaining.
+
+- When val([]int32) is given
+  - `var val []int32 = {1, 2}`
```

Comment:
how about boundary values, overflow, and negative value?

Reply:
boundary values, overflow are OK, but negative values is not needed because of cannot be set.

Reply:
I think it can be set because it is int32 not uint32 🤔
https://play.golang.org/p/3xfIIDU-D1G

Reply:
oh sorry, it's my mistake 😓 
I will add all patterns you commented.